docs/devel/writing-good-e2e-tests.md

```diff
@@ -0,0 +1,168 @@
+tests is a killer.  If under normal circumstances your pod enters the
+running state within 10 seconds, and 99.9% of the time within 30
+seconds, it would be gratuitous to wait 5 minutes for this to happen.
+Rather just fail after 30 seconds, with a clear error message as to
+why your test failed ("e.g. Pod x failed to become ready after 30
+seconds, it usually takes 10 seconds").  If you do have a truly
+legitimate reason for waiting longer than that, or writing a test
+which takes longer than 2 minutes to run, comment very clearly in the
+code why this is necessary, and label the test as "SLOW", so that it's
+easy to identify and avoid in test runs that are required to complete
+timeously (for example those that are run against every code
+submission before it is allowed to be merged).  1. Note that
+completing within, say, 2 minutes only when the test passes is not
+generally good enough.  Your test should also fail in a reasonable
+time.  We have seen tests that, for example, wait up to 10 minutes for
+each of several pods to become ready.  Under good conditions these
+tests might pass within a few seconds, but if the pods never become
+ready (e.g. due to a system regression) they take a very long time to
+fail and typically cause the entire test run to time out, so that no
+results are produced.  Again, this is a lot less useful than a test
+that fails reliably within a minute or two when the system is not
+working correctly.
+1. **Resilience to relatively rare, temporary infrastructure glitches
+or delays:** Remember that your test will be run many thousands of
+times, at different times of day and night, probably on different
+cloud providers, under different load conditions.  And often the
+underlying state of these systems is stored in eventually consistent
+data stores.  So, for example, if a resource creation request is
+theoretically asynchronous, even if you observe it to be practically
+synchronous most of the time, write your test to assume that it's
+asynchronous (e.g. make the "create" call, and poll or watch the
+resource until it's in the correct state before proceeding).
+Similarly, don't assume that API endpoints are 100% available.
+They're not.  Under high load conditions, API calls might temporarily
+fail or time-out. In such cases it's appropriate to back off and retry
+a few times before failing your test completely (in which case make
+the error message very clear about what happened, e.g. "Retried
+http://... 3 times - all failed with xxx".  Use the standard
+retry mechanisms provided in the libraries detailed below.
+
+## Some concrete tools at your disposal
+
+Obviously most of the above goals apply to many tests, not just yours.
+So we've developed a set of reusable test infrastructure, libraries
+and best practises to help you to do the right thing, or at least do
+the same thing as other tests, so that if that turns out to be the
+wrong thing, it can be fixed in one place, not hundreds, to be the
+right thing.
```

Comment:
Maybe add a pointer that (1) opening issues for specific better tooling is welcome, and (2) code implementing that tooling is even more welcome.